### PR TITLE
ARROW-11491: [Rust] support JSON schema inference for nested list and struct

### PR DESCRIPTION
**Pull request by houqp**
Not optimized for performance because if one needs performance, runtime schema inference shouldn't be used in the first place.

### CODE REVIEW

File: rust/arrow/test/data/mixed_arrays.json
Comment (houqp) [inline rust/arrow/test/data/mixed_arrays.json:4]: @nevi-me the I don't have the full context here, so my change of logic here might not be correct. Basically the current implementation throws out an type incompatible error when a column is detected to have both list and non-list value, for example: `"c": true"` v.s. `"c": [true, false]`.

Do you remember what's the reason for doing automatic conversion between scalar column to scalar list column? This behavior seems a bit too magical to me.

Comment (nevi-me) [reply rust/arrow/test/data/mixed_arrays.json:4]: There aim was to reduce the odds of throwing away data, converting the scalar to a single-item list.

I can't remember if I checked against Spark, but I'll do that during the day. I'm convinced that I've come across the behaviour elsewhere before, but can't remember where.

Comment (nevi-me) [reply rust/arrow/test/data/mixed_arrays.json:4]: What have you decided here; to preserve the old behaviour or change it?

**Review by nevi-me — APPROVED** [review]
The implementation looks good, JIRA should have been https://issues.apache.org/jira/browse/ARROW-10771, so I'll close that one as a duplicate

File: rust/arrow/test/data/mixed_arrays.json
Comment (houqp) [reply rust/arrow/test/data/mixed_arrays.json:4]: @nevi-me i just tested it with spark, looks like it's doing the conversion the other way, i.e. when a column contains both scalar and list values, it gets converted to string type. as a result, boolean lists are parsed into `"[true, false]"` string.

I don't have a strong opinion on this, but if we want to match spark, then we should probably go that fallback to string if incompatible types are detected. What do you think? @nevi-me @jorgecarleitao @andygrove .

Comment (alamb) [reply rust/arrow/test/data/mixed_arrays.json:4]: In general I think matching Spark's behavior is probably a good idea so as to try and maximize compatibility with the rest of the ecosystem.

That being said, if Spark's behavior is hard to use, confusing or unwieldy to implement, I think there is a reasonable case to be made for deviating from its behavior too. 

"it depends" lol

File: rust/arrow/src/json/reader.rs
Comment (alamb) [inline rust/arrow/src/json/reader.rs:1879]: What happened to `bb.value(9)`?

**Review by alamb — APPROVED** [review]
Bonus points if someone wrote down the rules this code is trying to implement (as a doc comment, eg. on `infer_json_schema_from_iterator `). I apologize if I have missed this somewhere. 

This code looks clean and easy to follow. I didn't review all the logic as I don't have a firm grasp on the rules this code is trying to implement (hence my request for documentation!) , so I feel ill qualified to try and check the implementation against the intention. 

That being said, the code looks good and well tested to me so I think it would be good to merge (and we can change it later if desired)

The only question I think should be answered / explained before this is merged is why the checked in file `test/data/mixed_arrays.json` ends up with one fewer column in the inferred schema after this PR.

File: rust/arrow/test/data/mixed_arrays.json
Comment (nevi-me) [reply rust/arrow/test/data/mixed_arrays.json:4]: @houqp @alamb one downside for now is that we don't have utilities to deal with JSON data once it's inferred and interpreted as strings (matching Spark behaviour). For example, we don't have anything to convert "[true, false]" back to an array of those values.

I suppose with schema inference we'll always have to make some compromises. I still prefer casting to list as that loses the least data, but if we use different behaviour, we could document the decision on the relevant function, so that other users could make changes in future (if they motivate for some other behaviour).

**Comment by alamb** [comment]
@houqp  -- I think this one is ready to go other than 

> The only question I think should be answered / explained before this is merged is why the checked in file test/data/mixed_arrays.json ends up with one fewer column in the inferred schema after this PR.

**Comment by houqp** [comment]
sorry for the delay, i have been busy lately. will update the PR this weekend to address all the feedback commented so far.

File: rust/arrow/src/json/reader.rs
Comment (houqp) [reply rust/arrow/src/json/reader.rs:1879]: b column is of list type, but the last row has a value of `"b":4`, so i replaced it with `"b": []` in the test data. I am going to revert my test data change to maintain the old behavior based on our previous discussions.

**Comment by houqp** [comment]
@alamb @nevi-me added more doc string to `infer_json_schema_from_iterator` to document the coercion rules. Ready for another round of review.

File: rust/arrow/test/data/mixed_arrays.json
Comment (houqp) [reply rust/arrow/test/data/mixed_arrays.json]: I have updated the PR to implement the old behavior so the inference rules are consistent with rules used in the actual value conversion code.

After thinking more on this, I think Spark's behavior would be a better design in the long run for the following two reasons:

* It preserves more info from the original JSON data. Once we convert scalar to array of scalar in Arrow, we will lose the information that some of the original values were just scalars. It's better to let users decide how they want to handle incompatible types in their query plans. For example, one might want to simply discard all rows with scalar values as a data cleaning step.
* Converting fields with incompatible types into string is unavoidable for more complex cases anyway. For example, when a field contains scalar, list and object type, there no other option other than reading the value as JSON string.

Perhaps we can refactor the code to match Spark's behavior after we added more JSON parsing kernels.

Comment (jorgecarleitao) [reply rust/arrow/test/data/mixed_arrays.json]: I think that spark's behavior is correct. Our JSON reader and schema inferer is in my opinion very broken for list arrays.

```
{"a": "hello"}
{"a": "world"}
{"a": "this"}
{"a": null}
{"a": "hello"}
{"a": "world"}
{"a": "this"}
{"a": null}
{"a": "hello"}
{"a": "world"}
{"a": "this"}
{"a": null}
{"a": ["a"]}
```

yields a `ListArray<String>` where each scalar is being interpreted as a one-element list. I can't see how this is reasonable.

@houqp I have code to handle arbitrary nested lists on my computer, but it requires that we do not make this kind of crazy stuff, i.e. it works as expected if no casting like the above is allowed.

File: rust/arrow/src/json/reader.rs
Comment (jorgecarleitao) [reply rust/arrow/src/json/reader.rs:61]: I agree. Much simpler!

**Review by jorgecarleitao — APPROVED** [review]
LGTM. Great work, @houqp !

File: rust/arrow/test/data/mixed_arrays.json
Comment (houqp) [reply rust/arrow/test/data/mixed_arrays.json]: > I have code to handle arbitrary nested lists

@jorgecarleitao could you expand a little bit more on this or share some of the code snippet? Don't we already support reading nested list values?

@nevi-me was mostly concerned with not able to operate on string fields that contains serialized JSON value, I wonder if your code can help alleviate that concern.

Comment (jorgecarleitao) [reply rust/arrow/test/data/mixed_arrays.json]: Sorry, scratch what I said. It is an unrelated issue that I mixed up.

File: rust/arrow/src/json/reader.rs
Comment (nevi-me) [inline rust/arrow/src/json/reader.rs:522]: This line (appears elsewhere) should no longer apply when following same behaviour as Spark

File: rust/arrow/test/data/mixed_arrays.json
Comment (nevi-me) [reply rust/arrow/test/data/mixed_arrays.json]: Is the conclusion then to keep the existing behaviour? I'm a bit confused 🙃

I went through the code, still a green light from me; so we can merge this

Comment (houqp) [reply rust/arrow/test/data/mixed_arrays.json]: Looks like it given Jorge already approved.

Comment (jorgecarleitao) [reply rust/arrow/test/data/mixed_arrays.json]: I do not think we need to/should address it in this PR: this is already a major improvement imo :)

**Comment by jorgecarleitao** [comment]
@alamb , was your question answered?

File: rust/arrow/src/json/reader.rs
Comment (alamb) [inline rust/arrow/src/json/reader.rs:520]: ❤️

**Review by alamb — APPROVED** [review]
I just re-checked @jorgecarleitao - my question is no longer relevant as the changes to the existing tests seems to have been removed from this PR. The documentation on `infer_json_schema_from_iterator` is also ery helpful. 

I re-reviewed the test cases in the PR and the behavior they exhibit look very reasonable to me. 

👍  nice work @houqp

**Pull request closed without merging**
